src/jit.c

```diff
@@ -0,0 +1,1564 @@
+            /* Use conditional move to avoid a branch. */
+            emit1(state, 0x48);
+            emit1(state, 0x0f);
+            emit1(state, 0x44);
+            emit1(state, 0xc1); /* cmove rax, rcx */
+        } else {
+            /* Restore dividend to RCX */
+            emit_pop(state, RCX);
+
+            /* Store the dividend in RAX if the divisor was zero. */
+            /* Use conditional move to avoid a branch. */
+            emit1(state, 0x48);
+            emit1(state, 0x0f);
+            emit1(state, 0x44);
+            emit1(state, 0xd1); /* cmove rdx, rcx */
+        }
+    }
+
+    if (dst != RDX) {
+        if (mod)
+            emit_mov(state, RDX, dst);
+        emit_pop(state, RDX);
+    }
+    if (dst != RAX) {
+        if (div || mul)
+            emit_mov(state, RAX, dst);
+        emit_pop(state, RAX);
+    }
+#elif defined(__aarch64__)
+    switch (opcode) {
+    case 0x28:
+        emit_dataproc_3source(state, false, DP3_MADD, dst, dst, src, RZ);
+        break;
+    case 0x2f:
+        emit_dataproc_3source(state, true, DP3_MADD, dst, dst, src, RZ);
+        break;
+    case 0x38:
+        divmod(state, JIT_OP_DIV_REG, dst, dst, src);
+        break;
+    case 0x98:
+        divmod(state, JIT_OP_MOD_REG, dst, dst, src);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+#endif
+
+#define SET_SIZE_BITS 10
+#define SET_SIZE (1 << SET_SIZE_BITS)
+#define SET_SLOTS_SIZE 32
+HASH_FUNC_IMPL(set_hash, SET_SIZE_BITS, 1 << SET_SIZE_BITS);
+
+/* The set consists of SET_SIZE buckets, with each bucket containing
+ * SET_SLOTS_SIZE slots.
+ */
+typedef struct {
+    uint32_t table[SET_SIZE][SET_SLOTS_SIZE];
+} set_t;
+
+/**
+ * set_reset - clear a set
+ * @set: a pointer points to target set
+ */
+static inline void set_reset(set_t *set)
+{
+    memset(set, 0, sizeof(set_t));
+}
+
+/**
+ * set_add - insert a new element into the set
+ * @set: a pointer points to target set
+ * @key: the key of the inserted entry
+ */
+static bool set_add(set_t *set, uint32_t key)
+{
+    const uint32_t index = set_hash(key);
+    uint8_t count = 0;
+    while (set->table[index][count]) {
+        if (set->table[index][count++] == key)
+            return false;
+    }
+
+    set->table[index][count] = key;
+    return true;
+}
+
+/**
+ * set_has - check whether the element exist in the set or not
+ * @set: a pointer points to target set
+ * @key: the key of the inserted entry
+ */
+static bool set_has(set_t *set, uint32_t key)
+{
+    const uint32_t index = set_hash(key);
+    for (uint8_t count = 0; set->table[index][count]; count++) {
+        if (set->table[index][count] == key)
+            return true;
+    }
+    return false;
+}
+
+#define UPDATE_PC(pc)                             \
+    emit_load_imm(state, RAX, (pc));              \
+    emit_store(state, S32, RAX, parameter_reg[0], \
+               offsetof(struct riscv_internal, PC));
```

Comment:
It is a bit confusing for mixing `PC` and `pc` in a macro. Can you clarify?

Reply:
This macro is redundant, we can remove it directly.